Changing usages of index(t) to a more robust isclose()

sim-eng-ii opened the pull request:
### Summary

This PR addresses issue #11 .
The PR replaces the usage of index(t) with a more robust approach using `np.isclose()` in the `plot_routines.py`

### Details and Comments

- Replaced instances of index(t) with a new function called `find_index_nearest_time_within_tolerance()`, utilising `np.isclose()`. This change ensures more accurate and reliable time comparisons.

- Introduced the `rtol` parameter to the `find_index_nearest_time_within_tolerance()` function, offering control over the relative precision used in time comparisons.

sim-eng-ii committed:
Update plot_routines.py with math.isclose()

sim-eng-ii committed:
set default value for t to 0.001

sim-eng-ii committed:
move math import before mpl_toolkits.axes_grid1

haggaila commented:
Hi @robotAstray thanks for this PR. 
The tests (which required some activation) show one lint error that is easy to fix (move an import statement).
In addition, please consider the following points:
The usage of `math.isclose()`, which seems to be incorrect according to the function's syntax.
The change will not achieve the same thing as the current code, which returns the index of the entry with time equal to `t` in the array by exact comparison. This should be improved to return the index the entry in the array with time closest to `t`.
It is also best not to set a default value for `t` because all values would be equally valid and there's no preferred one.
If you are interested, please submit the fixes.
Thanks again.

robotAstray commented:
@haggaila, thank you for the message, I will work on this and update you. Thank you.

sim-eng-ii committed:
Update plot_routines.py (#1)

Signed-off-by: robotAstray <robotastray@cantab.net>

robotAstray commented:
@haggaila I made changes to the logic, I welcome feedback, suggestions. Thank you. 
(I will add a docstring to the function shortly)

sim-eng-ii committed:
Add docstring to `find_nearest_index_t()` (#2)

Signed-off-by: robotAstray <robotastray@cantab.net>

sim-eng-ii committed:
rename new function to `find_index_nearest_time_within_tolerance()`

I renamed the function to something more appropriate that closely describe its behaviour.

Signed-off-by: robotAstray <robotastray@cantab.net>

sim-eng-ii committed:
format `plot_routines.py`

Signed-off-by: robotAstray <robotastray@cantab.net>

haggaila commented:
Thanks @robotAstray it looks better. I still see an issue - `find_index_nearest_time_within_tolerance()` may return `None` which will crash the invoking code. I'd say that the index of the closest entry should be returned, and in any case we would need to test the function to verify that it works as expected. Thanks.

sim-eng-ii committed:
Add comprehensive test suite for `find_index_nearest_time_within_tolerance()` (#3)

* `plot_routines.py`: return index of the closest entry in `find_index_nearest_time_within_tolerance()`

Signed-off-by: robotAstray <robotastray@cantab.net>

* fix docstring in plot_routines.py for `find_index_nearest_time_within_tolerance()`

Signed-off-by: robotAstray <robotastray@cantab.net>

* fix docstring in  plot_routines.py for `find_index_nearest_time_within_tolerance()`

Signed-off-by: robotAstray <robotastray@cantab.net>

* Create `/test/plot/__init__.py`

Signed-off-by: robotAstray <robotastray@cantab.net>

* Create test_plot_routines.py to test `find_index_nearest_time_within_tolerance()`

Signed-off-by: robotAstray <robotastray@cantab.net>

---------

Signed-off-by: robotAstray <robotastray@cantab.net>

sim-eng-ii committed:
fix formatting for /test/plot/__init__.py

Signed-off-by: robotAstray <robotastray@cantab.net>

sim-eng-ii committed:
remove trailing newlines in `/test/plot/__init__.py` (#4)

* remove __init__.py

Signed-off-by: robotAstray <robotastray@cantab.net>

* __init__.py formatted

Signed-off-by: robotAstray <robotastray@cantab.net>

---------

Signed-off-by: robotAstray <robotastray@cantab.net>

robotAstray commented:
Hi @haggaila, I've made some changes to the code based on your feedback. I've addressed the issue regarding the function returning `None` and added proper handling to ensure that the index of the closest entry is always returned. Additionally, I've added comprehensive tests to cover different scenarios and ensure the correct behaviour of `find_index_nearest_time_within_tolerance()`. The trailing newline issue has also been fixed in `test/plot/__init__.py`. Please let me know if there's anything else that needs attention. Thank you!
(I also think `test_plot_routines.py` may be used to add tests for other functions belonging to `plot_routines.py` I can work on it in another PR if needed)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/qiskit-community/lindbladmpo?pullRequest=23) <br/>All committers have signed the CLA.

sim-eng-ii commented:
> [![CLA assistant check](https://camo.githubusercontent.com/e4fc4d4e5d6bdfcb46625b6accf4d22d7c18986382c24830cba828a372b072cd/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/qiskit-community/lindbladmpo?pullRequest=23) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/qiskit-community/lindbladmpo?pullRequest=23) before we can accept your contribution.
> 
> **robotAstray** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/qiskit-community/lindbladmpo?pullRequest=23) it.

Thank you. I will sign it. Please note I changed my username. I will add the previous email address  to my account

sim-eng-ii commented:
@haggaila, I think this is completed unless you have any other suggestions. Thank you.
